### DISCUSSION TITLE
Exclude unstable updates from "group:allNonMajor"

### DISCUSSION THREAD

**Discussion by Okeanos** [renovatebot/renovate#19375] [Request Help] [answered]
Given a Renovate config like the following:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    ":separateMajorReleases",
    ":combinePatchMinorReleases",
    ":ignoreUnstable",
    "group:allNonMajor"
  ]
}
```

Renovate will group patches & minor updates into single PRs – which is awesome – however, these PRs also include updates for unstable dependencies based on the [rule](https://docs.renovatebot.com/configuration-options/#ignoreunstable) quoted below:

> By default, Renovate won't update any package versions to unstable versions (e.g. `4.0.0-rc3`) unless the current version has the same `major.minor.patch` and was already unstable (e.g. it was already on `4.0.0-rc2`). Renovate will also not "jump" unstable versions automatically, e.g. if you are on `4.0.0-rc2` and newer versions `4.0.0` and `4.1.0-alpha.1` exist then Renovate will update you to `4.0.0` only.

I personally would have expected Renovate to create separate PRs for these unstable dependencies and then apply the rules about how unstable updates are handled. Why? Because unstable version updates, in particular `0.x.y` versions such as `0.1.0` to `0.2.0` may contain breaking API changes (see [semver contract](https://semver.org/#spec-item-4)) that are by definition neither patch nor minor changes.

Including those unstable updates inside the `combinePatchMinorReleases` and `group:allNonMajor` PRs means that they break unexpectedly and need manual intervention.

Is there a way to exclude unstable (in particular `0.x.y` versions) from these groups? Am I overlooking a config rule?

**Comment by rarkins** [comment, marked as answer]
The grouping rule is correctly doing exactly what it says it does, which is group non-Major updates. You seem to be mixing up the concepts of major and _breaking_ (or potentially breaking). The rule is not called "group non-breaking".

You could achieve similar by replaceing this preset with your own package rule with same conditions, but add to it `matchCurrentValue` to exclude versions starting with `0` or including `-rc`.

**Reply by Okeanos** [reply]
I'll give that a try, thanks! 

The reason why I emphasised the semver definition regarding the `0.x.y` updates in particular is, that a `0.x.y` update is effectively a major upgrade according to semver because it may contain anything, even breaking changes. I personally would argue that the `allNonMajor` group preset should honor that definition but assuming your proposal works I can live with a workaround.

**Reply by Okeanos** [reply]
Using the following configuration works as expected, thanks a lot for the info and pointer :)

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    ":separateMajorReleases",
    ":combinePatchMinorReleases",
    ":ignoreUnstable"
  ],
  "packageRules": [
    {
      "groupName": "all non-major dependencies (exclude 0.x.y)",
      "groupSlug": "all-minor-patch-semver",
      "matchPackagePatterns": [
        "*"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch"
      ],
      "matchCurrentValue": "!/^v?0\\./"
    }
  ]
}
```
